src/Data/List/Permutation.agda

```diff
@@ -0,0 +1,134 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Permutation
+------------------------------------------------------------------------
+
+module Data.List.Permutation where
+
+open import Relation.Binary.PropositionalEquality as ≡
+  using (_≡_ ; refl ; setoid ; _≢_ ; isEquivalence)
+  renaming (trans to ≡-trans)
+open import Relation.Binary
+
+open import Data.List
+open import Data.List.Any
+
+open import Function
+
+open import Data.Empty
+open import Data.Product using (Σ ; _,_ ; proj₁ ; proj₂ ; _×_ ; ∃ ; ∃₂)
+
+open import Data.List.Properties
+
+module _ {a} {A : Set a} where
+
+  -- an inductive definition of permutation
+
+  data perm : Rel (List A) a where
+    unit  : ∀ {l} → perm l l
+    prep  : ∀ {l₁ l₂} x → perm l₁ l₂ → perm (x ∷ l₁) (x ∷ l₂)
+    swap  : ∀ {l₁ l₂} x y → perm l₁ l₂ → perm (x ∷ y ∷ l₁) (y ∷ x ∷ l₂)
+    trans : ∀ {l₁ l₂ l₃} → perm l₁ l₂ → perm l₂ l₃ → perm l₁ l₃
+
+  perm-sym : ∀ {l₁ l₂} → perm l₁ l₂ → perm l₂ l₁
+  perm-sym unit         = unit
+  perm-sym (prep x p)   = prep x (perm-sym p)
+  perm-sym (swap x y p) = swap y x (perm-sym p)
+  perm-sym (trans p p₁) = trans (perm-sym p₁) (perm-sym p)
+
+  perm-isEquivalence : IsEquivalence perm
+  perm-isEquivalence = record
+    { refl  = unit
+    ; sym   = perm-sym
+    ; trans = trans
+    }
+
+  perm-setoid : Setoid _ _
+  perm-setoid = record { isEquivalence = perm-isEquivalence }
+
+  -- A reasoning API to manipulate permutation based on the setoid we just defined.
+  -- for mperm, we can simply deligate the reasoning portion to ℕ, which is way richer.
+
+  module PReasoning where
+    open import Relation.Binary.EqReasoning perm-setoid
+      using (begin_ ; _∎ ; _≡⟨⟩_ ; _IsRelatedTo_ ; relTo)
+      renaming (_≈⟨_⟩_ to _p⟨_⟩_)
+      public
+
+    infixr 2 _∷_<⟨_⟩_  _∷_∷_<<⟨_⟩_
+
+    -- following APIs allow to "zoom in" to localize reasoning
+
+    _∷_<⟨_⟩_ : ∀ x (l : List A) {l₁ l₂ : List A} → perm l l₁ →
+                 (x ∷ l₁) IsRelatedTo l₂ → (x ∷ l) IsRelatedTo l₂
+    x ∷ l <⟨ r ⟩ rel = relTo (trans (prep x r) $ begin rel)
+
+    _∷_∷_<<⟨_⟩_ : ∀ x y (l : List A) {l₁ l₂ : List A} → perm l l₁ →
+                    (y ∷ x ∷ l₁) IsRelatedTo l₂ → (x ∷ y ∷ l) IsRelatedTo l₂
+    x ∷ y ∷ l <<⟨ r ⟩ rel = relTo (trans (swap x y r) $ begin rel)
+
+
+  -- properties for `perm`
+
+  perm-++ʳ : ∀ {l₁ l₂} l → perm l₁ l₂ → perm (l₁ ++ l) (l₂ ++ l)
```

Comment:
Also similarly, the word `perm` doesn't need to appear in any of these names, as the whole module is about that type.